drafts/str4d-stratum/draft1.rst

```diff
@@ -0,0 +1,256 @@
+  reconnecting.
+
+If ``client.reconnect`` is sent with empty parameters, the miner SHOULD
+reconnect to the same host and port it is currently connected to.
+
+``mining.suggest_target()``
+---------------------------
+
+Request (optional)::
+
+    {"id": 3, "method": "mining.suggest_target", "params": ["TARGET"]}\n
+
+- ``TARGET`` (hex): The target suggested by the miner for the next received job
+  and all subsequent jobs (until the next time this message is sent).
+
+The server SHOULD reply with ``mining.set_target``. The server MAY set the
+result id equal to the request id.
+
+
+Motivation
+==========
+
+Many existing cryptocurrency miners and pools use the original Stratum protocol
+for communication, in situations where the miner does not require any control
+over what they mine (for example, a miner connected to a local P2Pool node).
+However, the protocol is very specific to Bitcoin, in that it makes assumptions
+about the block header format, and the available nonce space. Zcash has made
+changes that invalidate these assumptions.
+
+Having a formal specification for a Zcash-compatible Stratum-style mining
+protocol means that existing pool operators and miner authors can quickly and
+easily migrate their frameworks to the Zcash network, with no ambiguity about
+interoperability.
+
+
+Rationale
+=========
+
+Why use the 256-bit target instead of a numerical difficulty?
+
+- There is no protocol ambiguity when using a target. A server can pick a
+  specific target (by whatever algorithm), and enforce it cleanly on submitted
+  jobs.
+
+  - A numerical difficulty must be converted into a target by miners, which adds
+    unnecessary complexity, results in a loss of precision, and leaves ambiguity
+    over the conversion and the validity of resulting submissions.
+
+- The minimum numerical difficulty in Bitcoin's Stratum protocol is 1, which
+  corresponds to ``powLimit``. This makes it harder to test miners and servers.
+  A target can represent difficulties lower than the minimum.
+
+Does a 256-bit target waste bandwidth?
+
+- The target is generally not set as often as solutions are submitted, so any
+  effect is minimal.
+
+- Zcash's proof-of-work, Equihash, is much slower than Bitcoin's, so any latency
+  caused by the size of the target is minimal compared to the overall solver
+  time.
+
+- For the current Equihash parameters (200/9), the Equihash solution dominates
+  bandwidth usage.
+
+
+Reference Implementation
+========================
+
+- `str4d's standalone miner`_
+
+.. _`str4d's standalone miner`: https://github.com/str4d/zcash/tree/standalone-miner
+
+
+Acknowledgements
+================
+
+Thanks to:
+
+- 5a1t for the initial brainstorming session.
+
+- Daira Hopwood for her input on API selection and design.
+
+- Marek Palatinus (slush) and his colleagues for their refinements, suggestions, and
+  robust discussion.
+
+This ZIP was edited by [TODO: fill in]
+
+
+References
+==========
+
```

Comment:
{"method":"mining.submit","id":4,"params":["1NqrZohQyRQ9J7FEbHyHTtDDVotaexeMxF","job_4bd76ced_0","fff67119","00000000000000000000000000000000000000000000000000000000","fd4005003e6fcca315ab42cb0041a36dfe7d218cab5dbfde0702215b93427d235d5241cf017860350acfae3285059bf9339a418800502bcaa54568fbb2de4bbb68829556b76bbe79d27fe8d4ba3ff15a8c1ed371bda30a0b6d622d8c0fced089bde81bc1e5fba6402cb397d42d646d867b697e45ff35e437a94b94aec1fbfa2e720e609c44c19652b7c98885d90de475d5d18c151dc116ec92662f0f462a8d596338b9240cc91a056e061a094d6b742fd75ffbef9068b603de7f662e4f52536f135e1999fc5a9b5b4c609495ad279f62345e165b951a945f60125396e90a9e1212f9fa8be0df82f07a2b51f86c5929a1cfeb2c81852a265523fd9b7e3dc11d0ce4bf6bcba366cfe09014966e76e69bbb8dded26f1a270155c69a5d96edd2d693a93df74e1e881b41f70e0539727d61345b873d41974841f810a4e5f1077713d59ae9048a096d6d9e08df4cd3a182af8dbf552900434191d4c247a22d67f783df7019ea1cbe98d82c38adaaa065ce60974087b39472cd119a44f0937e000ac05296e5c3c72b9e01554347ef9b9d88c62ed6422b965348813475fdd2af73cd2f9eb4c99559ffeae50a5b8cd3139313c49bbad5732ccdd3c5b400152abc322a4691d80eb56bc8efb61c723b0e5dec0f1ab7f42985aab980f355e1f274e3734cb0edeaa02fb59f7c959c2e9902b84643ce9c9bce2c62833f85bffdec790052680d848f88ac8c33c0bab507028ae1833dd77a0e76bdc7198e1f1f17651409eced283de763d2535e1a599f902b95fbb7b74771f692537fda1fc9be7968244669fb378a31167f3ab81451c4ee6f3c851b1ee8031581262a4e8c2b0488c36b7a551df9569bee708917b36d86e009e89940ba11e102c4e1afcce55f273512f51e27b59c4052f14e12f94a6d83d517591f1d9f1e2f397571e630a7f360447f272a6dab13931d3e2e01e286f717e6d6b184ef61b5c2bd6698c8795b21100900ed082295e947b868c3a5e0b34d658e9cdeb7351223410f32e512c99667923f00a02e44d6f3194a407daa8e6bd8a3b8eb746ad875f2dc76ee5ca1b7b0b204bbcb411398759904129544746e1815c09abbfa041120b60778552357ab61536b99e6cd10f6b60fc9c20550759f4217d504fca72271ddff8fb641b0bb5d9e0d3b34cb258639032611121fa56426f4ece01032d8133ae7b850361e01b9c7c4292ef373da9a29566638143d43e5bba95bbba14a55a30fedda71c23e5f930a246e3c8ed19e7059e27c83020f4d29459682cf28ee5e29f5664c649529d97268b2127bd42eca3ffc64c913c33c159d1018b37e0ef2e050ef781ddf0494c0c2229c816789d2dce9fc1a68c8f0e0f42e7d20bb058b1c54d65b53d076a35a5d964c65c8fbd236e3559eae30b33b01bfa87817bd1e8bad18ee03d3635d5d799704cea9a5bcea5c37ba1c1136746e1b652a3f79a84707924ad1ec255205dd35242793cb363d7ec5de9416079994a9ae22187d34f87295492eb4c8f91d7b836708a6ef0ce6960db3343b715ecdc30d14be71b95d74110b7d2c5b2d4113e29c921ca7757890d1321e260e19c5bf11682afab3816b719ed4ce723ca6fbd4c8cb2ab347b7afb443c7c93bf2e3266f6ea22b91f16355497ec7f29ff7b7c3f6d8775919454351e7a7f64092059d036ac5ba3b75db824245b4715b46409ff6ec1d05b29601861c3f53ee169681ec575dfa018092a0b8091ef40755d36ebd83e3038d9be52d96c48ab638e63499a31f9b163579741128feea7e07c7b717de3b440b6222b2fed3e3e1fbe5930390f18ef61040dc1da80e55b177fa9927b0d07b031ffec9ca51ffc0f509cf15171c1e2f1df9c90e92e3a9b9b3031e4ed99a72a81fce0be09a60c02d2eba83db5955291a13fd30a93b"]}

why the nonce is "0000000000000000000000000000000000000000000000000000000000000000"